[JENKINS-71438] Replace browser confirm with modal dialogs

mawinter69 opened the pull request:
# Dialogs
Add a new component for displaying dialogs in javascript as a replacement for the browser built-in dialogs and also have forms in dialogs.

See [JENKINS-71438](https://issues.jenkins.io/browse/JENKINS-71438).

### Supported methods:
- alert
- confirm
- prompt
- modal
- form

### Behaviour
Behaviour is as close as possible to the browser built in functions though there can't be a 1:1 replacement as the dialogs are executed asynchronously.
confirm, prompt and form return a Promise. The success is called when pressing
the OK button, the reject when pressing cancel or escape.

Forms can either submit the page or provide a FormData object to the success case of the Promise.

alert, confirm, prompt and form require to press the buttons, or use the Enter or Escape keys to close the dialog. A modal can be closed by clicking outside of the dialog.

### Usage
See https://github.com/jenkinsci/design-library-plugin/pull/264 for a working example
```
dialog.alert("Title");
dialog.alert("Title", {message: "Message"});

dialog.confirm("Are you sure", {message: "Do it"}).then(
  () => {
    doit();
  },
  ()=>{//This is the cancel case nothing todo
  }
);
  
dialog.prompt("Enter your name:").then(name => {
    console.log("Your name is: " + name);
  }, () => {}
);
```

### Additional changes
- This change also enables storing a message in the context menu, that is shown together with the title when deleting a project (or other places where a confirmation link is used). So instead of just having the browser confirm dialog that tells us we want to delete a project but not which one exactly, we now see the actual project.
- provides a `<l:delete message="" title="" permission="" />`  shortcut for calling the doDelete url on an item as confirmation link. 

### Internationalization
For commonly used texts for buttons a resource bundle is defined that is loaded via javascript.
These texts can also be used by plugins to avoid that they have to provide their own translations.

### Open points
- [x] allow to set the button color in <l:task or <l:confirmationLink
- [x] translation
- [x] form dialogs (with either submitting or returning the data as FormData object)
- [x] link confirmation dialogs
- [x] disable ok button when input of prompt is empty
- [x] Design Library
- [x] confirm used in user token and userseed
- [x] do not close dialog when clicking outside of dialog (=force using the buttons)
- [x] make the methods globally available so they can be easily used in plugins
- [x] need to escape displayname
- [x] bind enter to ok button
- [x] modal

### Where can we use it
- deleting things (done here at many places)
- Disconnecting agents
- Taking agents offline temporarily, update offline reason
- adding things (jobs, views, computers, clouds, users, ...)
- ...

Plugins where dialogs make sense:
- matrix-auth: adding user or group
- role-strategy: adding role (with pattern), adding a user/group
- agent-maintenance: deleting maintenance (single or selected),  adding new maintenance (already uses a dialog but not based on the `dialog` element of html).
- everywhere, where a browser built-in confirm, alert or prompt is used.
- ...

### Ideas
- Icons in the dialog either in the title or the message

### Screenshots:
![image](https://user-images.githubusercontent.com/17767050/236340593-5c28b080-c36d-4514-ac4b-ff3323cb3f6c.png)

![image](https://user-images.githubusercontent.com/17767050/236340701-834da936-2a70-4d58-895d-d812d08b35c0.png)

![image](https://user-images.githubusercontent.com/17767050/236340779-1c21f3e3-de9a-4ee0-8fe5-665b61e243b3.png)

![image](https://user-images.githubusercontent.com/17767050/236340867-34059fcc-8f5f-4a09-b93e-7f569dfcf713.png)

![image](https://user-images.githubusercontent.com/17767050/236340947-0f1bd7d5-1ee1-4ba5-806a-468290283439.png)



<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done
Tested interactively the places where confirmation links are used. 
Confirmation links already have a dedicated test, that is passed.
Besides that in the design library change linked above are more demos.

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- Replace browser confirm with modal dialogs at many places
- Developer: API for alert, confirm, prompt, modal and form dialogs

<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/
-->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7938"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

mawinter69 committed:
[JENKINS-13545] use dialogs instead of pages to delete various things

Use the confirm dialog approch for runs, computers, views and clouds

mawinter69 committed:
fix test and missing newline

mawinter69 committed:
change delete also for log recorders and users

mawinter69 committed:
modal with title for legend

The modal feature is displaying to complete content inside a div. For
the legend which is rather big the content must be scrolled. At present
this lead to the content being displayed without any margins and the
scrollbar arrows at top and bottom are partially overlapping with the
rounded corners of the modal.

This change introduces a dedicated title for the modal. The title will
be fixed on top of the modal and only the content is scrollable. A
proper margin at the bottom ensures that the scrollbar is not
overlapping with the corners.

mawinter69 committed:
make the title optional

reduce corner radius and remove the margin

mawinter69 committed:
native confirmation dialogs

instead of using the browser confirm method for confirmation use
a modal dialog

breaks ath

mawinter69 committed:
do not close when clicking outside

janfaracik commented:
Hey, looking good. I've had a WIP for a while now to do this - take a look if you need design ideas etc.

https://github.com/jenkinsci/jenkins/compare/master...janfaracik:jenkins:add-confirmation-modals?expand=1

mawinter69 commented:
Thanks for the info. I'm looking now into making this a general feature so we can replace all confirm, alerts and prompts easily, i.e. making the things blocking so you can just write, also for plugins
```
if (dialog.confirm("Do delete this?")) {
  delete()
}
```#
Still some way to go though

mawinter69 committed:
dialogs

add new component for displaying dialogs
Supported are
alerts
confirms
prompts

Behaviour is as close as possible to the browser built in functions.
confirm and prompt return a Promise. The resolve is called when pressing
the OK button, the reject when pressing cancel or escape.

Calling the methods:
dialog.alert("Message");
dialog.alert("Message", {title: "Title"});

dialog.confirm("Are you sure", {title: "Do it"}).then(() => {
    doit();
  }, ()=>{//This is the cancel case nothing todo
});

dialog.prompt("Enter your name:").then(name => {
    console.log("Your name is: " + name);
  }, () => {}
);

mawinter69 committed:
Merge branch 'master' into confirm

mawinter69 committed:
Merge branch 'master' into JENKINS-13545

mawinter69 committed:
Merge branch 'JENKINS-13545' into confirm

mawinter69 committed:
use dialog for plugin uninstallation

mawinter69 committed:
fix userseed

mawinter69 committed:
hide div of input

mawinter69 committed:
explicit merge of options

mawinter69 committed:
use Object.assign to merge defaults and options

mawinter69 commented:
Damn, the htmlunit doesn't support the showModal method of a dialog element in javascript

mawinter69 committed:
undo changes in modal/index.js

basil commented:
This repository is now using HtmlUnit 3.3.0, which supports modal dialogs, at HEAD.

mawinter69 committed:
Merge branch 'master' into confirm

mawinter69 committed:
fix tests

mawinter69 committed:
implement as class

support modal, e.g. for legend

mawinter69 committed:
rewrite without using class keyword

check input

mawinter69 committed:
disable animation

mawinter69 committed:
Merge branch 'master' into confirm

mawinter69 committed:
fine tuning

mawinter69 committed:
allow to load forms

mawinter69 committed:
styling

mawinter69 committed:
fix error with htmlunit

mawinter69 committed:
internationalization

In order to
provide resoure bundle for common verbs to be used in the buttons

mawinter69 commented on war/src/main/js/components/dialogs/index.js:
Should this be put into the modals component instead and renamed to window.modal?
The modal/index.js would be now obsolete

mawinter69 commented on war/src/main/scss/modules/dialogs.scss:
```diff
@@ -0,0 +1,83 @@
+  border-radius: 0.6rem;
+  border: none;
+  background: transparent;
+  box-shadow: var(--modal-box-shadow);
```
Still uses some definitions from the modals as they are already supported by the black theme.

mawinter69 commented on war/src/main/scss/modules/dialogs.scss:
```diff
@@ -0,0 +1,83 @@
+  border: none;
+  background: transparent;
+  box-shadow: var(--modal-box-shadow);
+  backdrop-filter: var(--modal-backdrop-filter);
```
defined in modals.scss

mawinter69 commented on war/src/main/scss/modules/dialogs.scss:
```diff
@@ -0,0 +1,83 @@
+  padding: 0;
+
+  &::backdrop {
+    background: var(--modal-backdrop-background);
```
defined in modals.scss

mawinter69 commented on war/src/main/scss/modules/dialogs.scss:
```diff
@@ -0,0 +1,83 @@
+
+  &__buttons {
+    padding: 0.5rem 1rem 1rem;
+    justify-content: center;
```
Buttons are centered in the dialogs. With OK button to the left. Is this ok or should they be put to the left or right side of the dialog and/or have the OK button on the right?

mawinter69 commented on war/src/main/scss/modules/dialogs.scss:
```diff
@@ -0,0 +1,83 @@
+  }
+}
+
+@keyframes jenkins-dialog-backdrop-animate-in {
```
There are currently no animate-out transitions. When I add them then htmlunit tests are not properly performing the page submit after clicking the ok button. On a live system everything works fine.

mawinter69 commented:
@jenkinsci/sig-ux

uhafner commented on war/src/main/scss/modules/dialogs.scss:
```diff
@@ -0,0 +1,83 @@
+$jenkins-dialog-padding: 1rem;
+
+.jenkins-dialog {
```
I wonder why our dialogs do not have a clear visual separation of the title and the button bar. Everything uses the same color and no line separates these elements. But maybe this is a matter of preference. I prefer the Boostrap styling in this case...

mawinter69 replied:
Something like this:
<img width="363" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/fdaeac91-0e87-4939-a4c1-5edb341e9eb4">

or this:
<img width="360" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/d59f85a1-094a-4a4d-8697-a57d1ba7a60d">

timja replied:
I'm not sold on the buttons in the middle currently, 

see:
https://uxdesign.cc/buttons-placement-and-order-bb1c4abadfcb

which recommends on the right

and on button ordering:
https://ux.stackexchange.com/questions/1072/ok-cancel-on-left-right

I think generally in Jenkins we have confirm on the left and cancel on the right (but may be wrong)

timja replied:
(I see you've moved them to the right in later commits than I was testing in the design library pull request)

What do you think about the button order?

uhafner replied:
> Something like this

Yes, I think such a styling will make the different areas more obvious.

mawinter69 replied:
What we have on config pages is that buttons are on the left and we have first `Save` then `Apply`
Maybe we should also make the button placement depend on the dialog type.
For `form` dialogs it might be better to have the buttons on the left with `Submit` being the first similar to the configure pages. This might also apply to the `prompt` dialog.
For other dialogs we can put them to the right and try to detect the OS and then put `OK` left (windows) or right (Mac, Linux).

uhafner replied:
I'm not a designer, maybe @janfaracik has an idea what looks better?

timja replied:
> What we have on config pages is that buttons are on the left and we have first `Save` then `Apply` Maybe we should also make the button placement depend on the dialog type. For `form` dialogs it might be better to have the buttons on the left with `Submit` being the first similar to the configure pages. This might also apply to the `prompt` dialog. For other dialogs we can put them to the right and try to detect the OS and then put `OK` left (windows) or right (Mac, Linux).

This sounds good to me.

mawinter69 replied:
applied now button placement that way.
Also applied now background color to title and buttons

timja replied:
This doesn't look right on dark theme at least:
<img width="342" alt="image" src="https://github.com/jenkinsci/jenkins/assets/21194782/5ebfaefe-0b53-4a85-a948-cbe595c05649">

On light it was pretty much invisible on my mac screen:
<img width="328" alt="image" src="https://github.com/jenkinsci/jenkins/assets/21194782/02b2d780-1481-4249-985b-0630aaf1459c">

Would a line work instead?

timja replied:
It feels weird the buttons moving all over the place depending on the dialog type, it may just be a problem in the design library as you see them all soon after each other and e.g. when you use the prompt one and alert pops up after but the buttons are on different side.

what do you think?

-----

Dark theme is looking good to me although the line seems a bit prominent on the default theme:
![image](https://github.com/jenkinsci/jenkins/assets/21194782/db22da68-a0e1-45cf-8bda-f7bfa4905299)

![image](https://github.com/jenkinsci/jenkins/assets/21194782/6cb53aae-d667-4282-8fc1-fb17ce22e3f3)

mawinter69 replied:
Actually there might be cases where you can see multiple dialogs with different type directly after each other. E.g. in matrix auth, plugin when you add a new user (uses a prompt) and that user is already added you would get an alert.

mawinter69 replied:
default theme with `var(--medium-grey)`

<img width="248" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/6341d4cc-a921-4a18-959b-aaf5142f8dbb">

mawinter69 committed:
cleanup properties

mawinter69 committed:
layout adjustements

define a minWidth of 300px. Otherwise dialogs with only short message
look quite bad

put buttons to the right in reverse order

mawinter69 commented on core/src/main/resources/hudson/model/ManageJenkinsAction/index.properties:
```diff
@@ -21,3 +21,4 @@
 # THE SOFTWARE.
 
 are.you.sure={0}: are you sure?
+sure=Are you sure?
```
removed

mawinter69 committed:
background and button placement

visual separation of title and buttons by applying a different
background color for them

place buttons ro the left for prompt and forms
detect os (windows or not) to decide button order

timja committed:
Lint:fix

timja commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/resources.js:
```diff
@@ -87,33 +87,38 @@ window.confirmAndRevokeAllSelected = function (button) {
+          selectedValues.push({ userId: userId, uuid: uuid });
+        }
+
+        // Workaround prototype.js breaking JSON.stringify
+        // Given this page is isolated, it's safe to not restore these after to the prototype version
```
looks like a bad merge
```suggestion
                // TODO simplify when Prototype.js is removed
```

timja commented on core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:
```diff
@@ -57,13 +57,14 @@ THE SOFTWARE.
               <j:choose>
                 <j:when test="${m.requiresConfirmation}">
-                  <l:confirmationLink href="${m.urlName}" post="${m.requiresPOST}" message="${%are.you.sure(m.displayName)}">
+                  <l:confirmationLink href="${m.urlName}" post="${m.requiresPOST}" message="${%sure}" title="${m.displayName}">
```
When I click this it jumps me to the top of the page which is a bit jarring. There's a flash as it happens.

mawinter69 replied:
I think the reason is that confirmationlink renders an `<a href="#" ...`
This makes the browser jump to the top. I think I've seen this behaviour even before.

mawinter69 replied:
https://stackoverflow.com/questions/2084750/javascript-anchor-avoid-scroll-to-top-on-click

timja commented on core/src/main/resources/hudson/model/User/sidepanel.jelly:
```diff
@@ -38,7 +38,7 @@ THE SOFTWARE.
       <t:actions actions="${it.transientActions}"/>
       <j:if test="${it.canDelete()}">
-        <l:task confirmationMessage="${%delete.user(it.displayName)}" href="${rootURL}/${it.url}/doDelete" icon="icon-edit-delete icon-md" title="${%Delete}" post="true" requiresConfirmation="true"/>
+        <l:delete urlPrefix="${rootURL}/${it.url}" title="${%Delete}" message="${%delete.user(it.displayName)}"/>
```
unsure where the change has happened but commenting here as related, if I click on the delete button from this page http://localhost:8080/jenkins/securityRealm/

I get redirected to:
http://localhost:8087/jenkins/securityRealm/user/john/delete

which is a 404 now:
<img width="1067" alt="image" src="https://github.com/jenkinsci/jenkins/assets/21194782/cf7913d8-0925-4e42-997d-24b4901b1e73">

mawinter69 replied:
That was a bug that was fixed with https://github.com/jenkinsci/jenkins/pull/8131

timja commented on core/src/main/resources/lib/layout/delete.jelly:
```diff
@@ -0,0 +1,82 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout">
+  <st:documentation> <![CDATA[
+    Adds a delete link in the left navigation bar of Jenkins.
```
from what I can tell it's done like this:
```suggestion
    Adds a delete link in the left navigation bar of Jenkins.
    
    @since TODO
```

timja reviewed (commented):
Looking really good I've tested it quite thoroughly locally.

some issues I found

timja committed:
Merge branch 'master' into confirm

mawinter69 committed:
Update core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/resources.js

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

mawinter69 committed:
use lines to separate header and buttons

mawinter69 committed:
prevent default in confirmation link

the default makes the browser jump to the top of the page

mawinter69 committed:
update delete jelly description

mawinter69 committed:
remove modals script and css

functionality is moved to dialogs

timja committed:
Update styles.scss

mawinter69 committed:
change separator line color

mawinter69 committed:
always place buttons to the right

only differentiate between windows and unix/mac

mawinter69 committed:
Merge branch 'master' into confirm

Kevin-CB commented:
The security review is currently in progress.

Kevin-CB reviewed (approved):
I have tested your PR locally and from a security perspective, I don't see any blockers.

mawinter69 committed:
Merge branch 'master' into confirm

timja commented:
ATH is failing https://github.com/jenkinsci/acceptance-test-harness/pull/1256

janfaracik commented:
Looking great!

Couple of things - 

I'm not sure if the borders benefit the dialog? If there was a lot of content I could see it being useful but right now I think it divides the dialog unnecessarily. 

I'd also make everything a touch chunkier, currently theres varying font sizes and fairly tight padding - making everything a little larger makes it feel more consistent and also makes the hit targets larger, e.g.

<img width="540" alt="image" src="https://github.com/jenkinsci/jenkins/assets/43062514/5f07a6a7-3616-4735-a7a5-79b502adf31f">

The code for the above is 

<details>
<summary>Click to expand</summary>
$jenkins-dialog-padding: 1.3rem;

.jenkins-dialog {
  border-radius: 1rem;
  border: none;
  background: transparent;
  box-shadow: var(--dialog-box-shadow);
  backdrop-filter: var(--dialog-backdrop-filter);
  animation: jenkins-dialog-animate-in 0.25s cubic-bezier(0, 0.68, 0.5, 1.5);
  overflow: hidden;
  padding: 0;

  &::backdrop {
    background: var(--dialog-backdrop-background);
    animation: jenkins-dialog-backdrop-animate-in 0.15s;
  }

  &__contents {
    overflow-y: auto;
    overflow-wrap: break-word;
    padding: 0 $jenkins-dialog-padding;
    max-height: 75vh;
    font-size: 1rem;
    color: var(--text-color-secondary);
  }

  &__input {
    display: flex;
    padding: 0 1rem 1rem;
  }

  &__buttons {
    padding: $jenkins-dialog-padding;

    .jenkins-button {
      font-size: 0.9rem;
      padding: 0.7rem 0.6rem;
      min-width: 100px;
    }

    &--right-windows {
      justify-content: right;
    }

    &--right-unix {
      justify-content: right;
      flex-direction: row-reverse;
    }
  }

  &__title {
    font-size: 1.1rem;
    font-weight: 600;
    padding: $jenkins-dialog-padding;
  }

  &__subtitle {
    font-size: 1rem;
    font-weight: 600;
    color: var(--text-color-secondary);
    padding: 0;
    margin: 0 0 1rem;
  }

  &__close-button {
    position: absolute;
    top: $jenkins-dialog-padding - 0.5rem;
    right: $jenkins-dialog-padding - 0.5rem;
    aspect-ratio: 1;
    padding: 0;

    &::before,
    &::after {
      border-radius: 100%;
      outline: none;
    }

    &::after {
      backdrop-filter: contrast(0.9) blur(5px);
    }
  }
}

@keyframes jenkins-dialog-backdrop-animate-in {
  from {
    opacity: 0;
  }
}

@keyframes jenkins-dialog-animate-in {
  from {
    scale: 85%;
    opacity: 0;
  }
}
</details>

---

Also unsure about reordering the buttons depending on OS? I might be alone here but I use Jenkins across a variety of operating systems so having the controls change order would be a little confusing for muscle memory. It also looks like the flex box `flex-direction` doesn't alt + tab correctly.

mawinter69 committed:
Merge branch 'master' into confirm

mawinter69 commented:
Making the message text with secondary color has the drawback that this doesn't look so good when there is no title. One could make the color depending if a title is present or not.
Which button layout would you propose?
- Buttons on the left (OK button left)
- Buttons on the right windows style (OK button left)
- Buttons on the right unixstyle (OK button right)

mawinter69 commented:
> I'd also make everything a touch chunkier, currently theres varying font sizes and fairly tight padding - making everything a little larger makes it feel more consistent and also makes the hit targets larger, e.g.
> 

The proposed change has the problem that when there is no title that then the padding is missing, also for prompts it is not working, the text is directly above the input without any padding then. Probably I need to make the padding depend on title is there or not

janfaracik commented:
> Making the message text with secondary color has the drawback that this doesn't look so good when there is no title. 

Is there ever a situation where the dialog would have a description but not a title?

> Which button layout would you propose?

I'd go for 'Buttons on the right unixstyle (OK button right)' personally

> also for prompts it is not working, the text is directly above the input without any padding then. Probably I need to make the padding depend on title is there or not

If you made the dialog a flex box, you could use the `gap` attribute to set a gap between each element, you wouldn't need the padding then :)

mawinter69 commented:
> > Making the message text with secondary color has the drawback that this doesn't look so good when there is no title.
> 
> Is there ever a situation where the dialog would have a description but not a title?

I would not exclude that possibility, it is not always necessary to have both.
Instead of having the first parameter be the message I could make it the title for alert, confirm and prompt
> 
> > Which button layout would you propose?
> 
> I'd go for 'Buttons on the right unixstyle (OK button right)' personally

Maybe we need to do a poll. 
As the OK button has different color it already draws attention to it.
> 
> > also for prompts it is not working, the text is directly above the input without any padding then. Probably I need to make the padding depend on title is there or not
> 
> If you made the dialog a flex box, you could use the `gap` attribute to set a gap between each element, you wouldn't need the padding then :)

will try that out

mawinter69 committed:
Merge branch 'master' into confirm

mawinter69 committed:
flexbox

use flexbox for the dialog
add titles for some confirms around tokens and seeds

mawinter69 committed:
apply prettier

mawinter69 commented:
@janfaracik I'm using now a flex box, made it a bit easier

buttons are now always unix style

The first parameter to alert, confirm and prompt will now be the title and not the message.

Adjusted some occurences of confirms to have both title and message

mawinter69 committed:
Merge branch 'master' into confirm

timja reviewed (approved):
Looks great!

janfaracik commented:
Looking great! Couple non-blockers from me -

1. There isn't a closing animation (there is for the 'Icon legend' modal)
2. A higher min width would look good, it's a little too narrow in the below example 

<img width="747" alt="image" src="https://github.com/jenkinsci/jenkins/assets/43062514/a166ca01-e472-486d-8199-ccde2ecf2d2e">

Something like 450px makes it look better:

<img width="684" alt="image" src="https://github.com/jenkinsci/jenkins/assets/43062514/c8e24980-2285-469c-abec-3b1451ecc5d5">

mawinter69 committed:
Merge branch 'master' into confirm

mawinter69 committed:
increase minwidth

mawinter69 commented:
> 2. A higher min width would look good, it's a little too narrow in the below example

increased minwidth to 450px

timja commented:
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

MarkEWaite commented:
Congratulations on the merge of the pull request.  I experimented with it shortly after its release and like the result very much.  It provides a much better user experience than the previous browser delete dialog.

I need your help with the plugin bill of materials to adapt to this pull request.  The Jenkins plugin bill of materials is showing new test failures in the config-file-provider-plugin with Jenkins 2.415.  When I bisect the test failures in Jenkins core, the first commit that shows the test failure is this one.

The test failures are listed in the plugin BOM test results at https://github.com/jenkinsci/bom/runs/15255225821 and are available on ci.jenkins.io at https://ci.jenkins.io/job/Tools/job/bom/job/master/1899/testReport/

@mawinter69 would you be willing to review the 4 test failures and their source code and suggest changes to the config file provider plugin to fix the issue?

To see the issue, I ran the following maven command with a checkout of the master branch of the config file provider plugin:

```
mvn clean -DforkCount=1 -Djenkins.version=2.415 -Dtest=ConfigFilesSEC1253Test.xssPrevention,ConfigFilesSEC1253Test#regularCaseStillWorking,FolderConfigFileActionTest#xssPreventionInFolder,Security2002Test#xssPrevention verify
```

To confirm that 2.414 did not have the issue, I ran the command:

```
mvn clean -DforkCount=1 -Djenkins.version=2.414 -Dtest=ConfigFilesSEC1253Test.xssPrevention,ConfigFilesSEC1253Test#regularCaseStillWorking,FolderConfigFileActionTest#xssPreventionInFolder,Security2002Test#xssPrevention verify
```

MarkEWaite commented:
@mawinter69 it looks like this change has introduced a regression in the pop-up menu for "Manage Jenkins" that is available when clicking the down arrow of the "Dashboard" link at the top of the page.  More details are described in [JENKINS-71744](https://issues.jenkins.io/browse/JENKINS-71744).

Would you be willing to review it and see what can be done to resolve it?

mawinter69 commented:
I'll take a look
